core/raftlease/fsm.go

```diff
@@ -387,20 +387,26 @@ func (r *response) Notify(target NotifyTarget) {
 	}
 }
 
+// Apply log is invoked once a log entry is committed.
+// It returns a value which will be made available in the
+// ApplyFuture returned by Raft.Apply method if that
+// method was called on the same Raft node as the FSM.
 // Apply is part of raft.FSM.
 func (f *FSM) Apply(log *raft.Log) interface{} {
-	var command Command
-	err := yaml.Unmarshal(log.Data, &command)
+	command, err := unmarshalCommand(log)
 	if err != nil {
 		return &response{err: errors.Trace(err)}
 	}
-	if err := command.Validate(); err != nil {
-		return &response{err: errors.Trace(err)}
-	}
 
 	f.mu.Lock()
 	defer f.mu.Unlock()
 
+	return f.apply(command)
+}
+
+// apply extacts out of the command operation invocations, so that the batching
+// FSM can make use of the same logic.
+func (f *FSM) apply(command Command) *response {
```

Comment:
Can you please add a note here that the method expects the caller to be holding the lock